src/org/rascalmpl/library/util/ErrorRecovery.java

```diff
@@ -169,4 +171,47 @@ private ScoredTree disambiguateAmb(ITree amb, boolean allowAmbiguity, Map<IConst
 
         return result;
     }
+
+    public IList findAllErrors(IConstructor tree) {
+        IListWriter errors = rascalValues.listWriter();
+        collectErrors((ITree) tree, errors, new HashSet<>());
+        return errors.done();
+    }
+
+    private void collectErrors(ITree tree, IListWriter errors, Set<IConstructor> processedTrees) {
+        Type type = tree.getConstructorType();
+
+        if (type == RascalValueFactory.Tree_Appl) {
+            collectApplErrors(tree, errors, processedTrees);
+        } else if (type == RascalValueFactory.Tree_Amb) {
+            collectAmbErrors(tree, errors, processedTrees);
+        }
+    }
+
+    private void collectApplErrors(ITree appl, IListWriter errors, Set<IConstructor> processedTrees) {
+        if (processedTrees.contains(appl)) {
+            return;
+        }
+        processedTrees.add(appl);
```

Comment:
performance wise we can skip 1 iteration of the hash-tree by doing:

```java
if (!processedTrees.add(appl)) {
    return;
}
```

Reply:
I have already completed this PR, I will push this change directly to the `feat/error-recovery` branch.